### ISSUE TITLE
Dart 2.0 Spec

### ISSUE THREAD

**Issue by f2prateek** [f2prateek/dart#40]
* [x] Split compiler and API artifacts
* [x] Use JavaPoet
* [ ] Add a Gradle plugin module  that Automagically™ sets up the compiler and API artifacts as dependencies with the right scope (see [Hugo](https://github.com/JakeWharton/hugo/tree/master/hugo-plugin) for example)
* [x] IntentBuilder DSL (https://cloudup.com/cdZ8PWTSiEH)

These might be done before v2 lands:
* [x] An intent builder that works on top of the generated bundler
* [ ] A fragment builder that works on top of the generated bundler

**Comment by f2prateek** [comment]
We might also consider splitting the bundler code into a third module so that not all users need to have to carry the burden of the bundler generated code.

**Comment by stephanenicolas** [comment]
What do you mean by bundler generated code ?
Can we contribute ?  Is there a branch already set for v2 ?

**Comment by stephanenicolas** [comment]
What do you mean by bundler generated code ?
Can we contribute ?  Is there a branch already set for v2 ?

**Comment by stephanenicolas** [comment]
Sorry for duplicate. The phone version is limited..
Also, what would be the name of the annotation for fragment arguments ?

**Comment by stephanenicolas** [comment]
And for the intent builder, should we target a dsl ?

**Comment by f2prateek** [comment]
@stephanenicolas sorry for the delay. The bundler is basically just the intent builder but a more generic version - it generates code to build a `Bundle` instead of an `Intent` - https://github.com/f2prateek/dart/pull/39

We should have the DSL as well - Added that to the original comment.

**Comment by stephanenicolas** [comment]
@f2prateek , I am actually on a couple of project, one of pure research and one with a lot of research also at the job. But I am interested in contributing and I think it should not be that long.

So, let me know when you wanna start the development of v2.0. I contacted you via hangout, so you can ping me when you are ready.

**Comment by f2prateek** [comment]
I think we still need some more discussion before we start implementing new features. 

Once we can finalize the v2 spec, I can add you as a contributor and we can start working on it :)

**Comment by stephanenicolas** [comment]
Hi Pratek,

I can contribute to dart tomorrow. Probably early. If you are free, you can
reach me on hangout and we can talk about it.

Stephane
Le 2015-07-11 12:09, "Prateek Srivastava" <notifications@github.com> a
écrit :

**Comment by stephanenicolas** [comment]
@f2prateek I updated master. 

We now have 
* artifact split for runtime and compile time, using provided scope as needed
* intent builder factory, a.k.a Henson 
* a DSL for creating intents that is generated automatically as discussed in issue #33 
* all related issues are closed, one is pending

Can you do a code review, to my mind there are still a few things to be done before releasing, mainly getting code coverage and increase it to an acceptable level (>95%..though this can be lowered to release earlier). 
Also, I am still considering masking the intent builder with a generated caller like 
```java
Henson.goto(SimpleActivity).dsl().build();
```
to prevent devs from having to know about the location of the intent builder.

I think the current status of the code would be good enough for a release candidate.

**Comment by stephanenicolas** [comment]
Btw, you mentioned, you wanted to achieve a gradle plugin. I am not so sure it's essential, but why not. Though, I am not sure how you want to do that since our project is under maven. I tried to create a gradle plugin under maven, but it looks like the wrong tool for the task. How did you want to do that ?

**Comment by f2prateek** [comment]
We can add it later, not a pressing need for v2.

**Comment by stephanenicolas** [comment]
I created Henson, the generated navigator class. So the API is complete. I still have a few tests to add and then maybe we can agree on some release schedule and requirements. What do you think @f2prateek  ?

**Comment by f2prateek** [comment]
Sure, works for me. Let me know when I should cut a RC release.

**Comment by stephanenicolas** [comment]
Yop, it's ready now. I have added documentation, tests, polished the APIs. 
The code is in a clean state for a RC 1 IMHO.

If you agree we should create a ticket to get CI running code coverage. I think it's the only thing that remains to be done before an official release (and ensuring we got a coverage we are not afraid to publish ;), but I guess we are high already, the code is well tested).

Thanks a lot for letting me be a contributor of Dart. I don't know how we can advertise that, but it would be nice to have some credits for Groupon, I did all the work on Dart, and @Macarse too, during our office hours @ Groupon. Maybe just a little word about it when we make the official release announcements.

**Comment by f2prateek** [comment]
@stephanenicolas absolutely, we can definitely add a note to the README and 2.0.0-RC1 changelog, and I'll make sure to give you credit for the 2.0.0 release.

The only thing blocking the RC release is the CI failure on Java 7 - https://travis-ci.org/f2prateek/dart/jobs/71438339. 

I haven't used any code coverage tools with Maven, but if you have any ideas feel free to add them to the repo.

If possible, send pull requests instead of pushing to master directly so the CI can catch errors before code gets merged in, and we avoid publishing potentially broken snapshot builds.

**Comment by f2prateek** [comment]
Thanks for fixing the CI, I'll make the RC release today.

**Comment by stephanenicolas** [comment]
Hi @f2prateek Could you do the RC finally ?
I can't access search.maven.org right now..

**Comment by f2prateek** [comment]
@stephanenicolas Sorry I did try on Friday but forgot to tell you. The problem was that the javadoc generation fails:

https://cloudup.com/cs8e_rQq1eC

I think it's because of the new Java 8 javadoc linting. Let me know if you want to fix those or I can take a deeper look.

**Comment by f2prateek** [comment]
run `mvn javadoc:javadoc` on Java 8 to reproduce.

**Comment by stephanenicolas** [comment]
@f2prateek I would appreciate if you could fix them and release, I wanna integrate Dart 2.0 in our app tomorrow. 

Thx in advance,
 S.

**Comment by f2prateek** [comment]
Published the release, should be available in a couple of minutes, and show up on Maven search in a couple of hours.

**Comment by stephanenicolas** [comment]
@f2prateek 
Dart 2 is gonna get mentioned in the next issue of Android Weekly this week end. It might be time for a release candidate 2 ;)

**Comment by f2prateek** [comment]
@stephanenicolas is there anything else that needs to go into RC2 or can I publish what's on master?

Also do you have a sonatype account? If you can share the username, I can ask them for you to have access to publish releases.

**Comment by stephanenicolas** [comment]
Hey @f2prateek. My sonatype login is robospice. I publish many artifacts
from many projects with this user name, not only robospice.

For me, everything is good for a RC 2, sorry to have missed your mail. We
can talk about the remaining issues before RC 3, and hopefully get some
feedback from the community before releasing an official Dart 2 artifact.

S.


2015-07-25 17:22 GMT-07:00 Prateek Srivastava <notifications@github.com>:

**Comment by f2prateek** [comment]
Requested access https://issues.sonatype.org/browse/OSSRH-16768

**Comment by stephanenicolas** [comment]
Dart 2.0.0-RC1 has been released on sonatype. Should hit central in less than 4 hours.

**Comment by f2prateek** [comment]
It'll actually be up on central fairly quickly (~10 minutes) - just takes longer to show up on search.

**Comment by stephanenicolas** [comment]
Good to know.. I understand your last release message now. Thx

2015-07-27 17:54 GMT-07:00 Prateek Srivastava <notifications@github.com>:

**Comment by stephanenicolas** [comment]
@f2prateek we aim to release DART 2 this week, with @dlemures .

**Comment by f2prateek** [comment]
Awesome!

**Comment by stephanenicolas** [comment]
@f2prateek  

We are ready to release tomorrow. @dlemures will. We need to update the docs, announce on social networks + android dev... and that's it :)

@dlemures has been doing a great job for release 2 of Dart, @johncarl81 helped us a lot too. The Parceler support is really neat, and everything is well tested.

**Comment by johncarl81** [comment]
:+1:

**Comment by stephanenicolas** [comment]
Release check list for dart 2 : 

[x] make dart 2 ready
[ ] upgrade version number 
[ ] release on CLI
[ ] go to sonatype web interface
[ ] check + close + release 
[ ] issue MGA branch with dart 2 
[ ] polish docs + say to prateek
[ ] announce on Android Dev Weekly + google + + twitter

**Comment by stephanenicolas** [comment]
@f2prateek @johncarl81 Dart 2.0.0 is shipped ! 
Thx guys for your effort in this release. @dlemures great job ! :D

**Comment by dlemures** [comment]
:+1: :+1: :+1:

**Comment by stephanenicolas** [comment]
Everything done. Dart 2 is gonna be mentioned in Android Dev Weekly #196. I close the issue.

**Comment by f2prateek** [comment]
Excellent, awesome work everyone!
